### ISSUE TITLE
Would be great to have realm.empty(Model.class)

### ISSUE THREAD

**Issue by RobertoArtiles** [realm/realm-java#1862]
Hi team,

Would be great to be able to generate empty RealmResults<Model> collection for the testing purposes, and for following NonNull result contract when you wrap realm DB fetches in your own methods. Example:

```java
@NonNull
public static RealmResults<Foo> findAllFoos(Realm realm, boolean bar) {
    if (bar) {
        return realm.where(Foo.class).findAll();
    } else {
        return realm.empty(Foo.class); // to comply with non-null contract
    }
}
```

**Comment by zaki50** [comment]
Hi @RobertoArtiles 
Thank you for your request.

We had some similar requests before.
Current workaround is to make a `RealmQuery` which does not match any existing object.
It's not efficient way though...

I prefer `realm.where(Foo.class).findNone()` or similar to `realm.empty(Foo.class)` if we add this functionality.

**Comment by cmelchior** [comment]
Perhaps a factory method on RealmResults: `RealmResults.empty()`?

**Comment by zaki50** [comment]
@cmelchior It is confusable with `RealmResults.clear()`, isn't it?

**Comment by cmelchior** [comment]
Maybe, but I am talking about a static method, clear can only be called on an instance:
```
RealmResults.empty()
results.clear()
```
I was going for something like `Collections.emptyList` (which would also work depending on your expecting a RealmResults or a List)

**Comment by zaki50** [comment]
If it's acceptable for `RealmResults.isValid()` to always return `true`, it seems to work and is easy to implement.

**Comment by zaki50** [comment]
Oh, `RealmResults` has `where()` and it also does not work correctly.

**Comment by cmelchior** [comment]
IMO it would make sense for an empty RealmResult set to be a valid RealmResults, at least if you specifically asked for the empty set

**Comment by RobertoArtiles** [comment]
>>Current workaround is to make a RealmQuery which does not match any existing object.
Yes, I used it for my testing purposes. But yeah, would be great to have some special method for that.

Also, I changed my initial request by leaving only the "testing purpose". Such thing should not be used to comply with the non-null contract in real code, because it might lead to unexpected behaviour: when you wait for the result set to change, but it won't because it doesn't have the proper query in it.

**Comment by zaki50** [comment]
My previous comment is wrong.
`results.where()` does not return objects which is not included in original `results`.
So, any `RealmQuery` created from empty `RealmResults` always be empty.

**Comment by slidenerd** [comment]
I was about to open a new issue but searched for this and thought I would add my comment here as well.

What would be the best way to achieve the following currently with Realm

I first query for empty RealmResults, lets say I am depending on a value that I need to retrieve from the network and based that value, my further queries are based

final RealmResults<Test> results = realm.where(Test.class).equals("id", NON_EXISTENT_VALUE).findAll();

results.addChangeListener(new RealmChangeListener(){
public void onChange(){
//...

results.removeChangeListener(this);
}
})

Now at some point down the line, I get the real results

results = realm.where(Test.class).equals("id",KEY).findAll();

How do I manage change listeners in such a scenario and what is the best way to do this, I have a situation currently where I want to load posts and I need the user id for whom I need to load posts, I also need to display an empty results before I load the user id from the network, any suggestions?

**Comment by vietstone** [comment]
Any update on this?

**Comment by vietstone** [comment]
@beeender Could you provide a demo? As I'm as Swift guy, so I don't understand Java code much :)

**Comment by beeender** [comment]
A demo for what? since #3834 has not been merged, it is difficult to make a demo to produce a empty results based on current master code.

**Comment by vietstone** [comment]
oh, got it. Thank you @beeender

**Comment by ShaharHD** [comment]
@beeender Is this something still considered?

**Comment by beeender** [comment]
@ShaharHD Yes, it is! But it is not a high priority feature for us to implement right now.

**Comment by cmelchior** [comment]
I did look briefly into this because it would also make the `@LinkingObjects` API much nicer to use:

However, queries is turning out to be quite the headache
```
@LinkingObjects("dog")
private final RealmResults<Person> owners = RealmResults.empty();

// This would crash horribly
owners.where().equalTo("name", "john").findAll(); 
``` 
We could of course just throw if calling `RealmResults.empty().where()` but that would be rather ugly.

**Comment by Zhuinden** [comment]
You can now do it with `realm.where(clazz).isFalse().alwaysFalse().findAll()`

**Comment by astamato** [comment]
Anyone still around here in 2019? Feature would still be amazing to have!
